src/cpp/server/orca/orca_service.cc

```diff
@@ -0,0 +1,224 @@
+//
+// Copyright 2022 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "google/protobuf/duration.upb.h"
+#include "upb/upb.hpp"
+#include "xds/data/orca/v3/orca_load_report.upb.h"
+#include "xds/service/orca/v3/orca.upb.h"
+
+#include <grpcpp/ext/orca_service.h>
+#include <grpcpp/impl/codegen/server_callback_handlers.h>
+
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/iomgr/timer.h"
+
+namespace grpc {
+namespace experimental {
+
+//
+// OrcaService::Reactor
+//
+
+class OrcaService::Reactor : public ServerWriteReactor<ByteBuffer>,
+                             public grpc_core::RefCounted<Reactor> {
+ public:
+  explicit Reactor(OrcaService* service, const ByteBuffer* request_buffer)
+      : service_(service) {
+    GRPC_CLOSURE_INIT(&on_timer_, OnTimer, this, nullptr);
+    // Get slice from request.
+    Slice slice;
+    request_buffer->DumpToSingleSlice(&slice);
+    // Parse request proto.
+    upb::Arena arena;
+    xds_service_orca_v3_OrcaLoadReportRequest* request =
+        xds_service_orca_v3_OrcaLoadReportRequest_parse(
+            reinterpret_cast<const char*>(slice.begin()), slice.size(),
+            arena.ptr());
+    if (request == nullptr) {
+      Finish(Status(StatusCode::UNKNOWN, "could not parse request proto"));
```

Comment:
wouldn't internal be a better error code?

Reply:
Good catch.  Looks like we use INTERNAL for the public C++ API, so we should do the same thing here.

Done.